[export] simplify automatic dynamic shapes processing

pianpwk opened the pull request:
Removing `_transform_shapes_for_default_dynamic` and `assume_static_by_default=False` as added in https://github.com/pytorch/pytorch/pull/133620.

This reverts back to `assume_static_by_default=True` with the use of dynamo decorators (e.g. `maybe_mark_dynamic, mark_static`, instead) for handling Dim.AUTO & Dim.STATIC instead. This is easier to maintain, as it doesn't requiring reasoning about "inverting" the dynamic_shapes specs, and also opens up usage of other decorators (`mark_dynamic, mark_unbacked`).

On the user side this change has no effect, but internally this means dynamic behavior is determined only by the `dynamic_shapes` specs (ignoring user-side input decorators following https://github.com/pytorch/pytorch/pull/135536), but transferring this information for _DimHints via decorators, for Dynamo/non-strict to create symbolic_contexts accordingly, e.g. https://github.com/pytorch/pytorch/blob/7c6d543a5b3d65d8f49420e60cda150faaa5b8a0/torch/_dynamo/variables/builder.py#L2646-L2666

One caveat is we don't raise errors for dynamic decorators on the user side, since we don't know if they're from user markings, or from re-exporting with inputs we've previously marked. 

Differential Revision: D63358628

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -94,8 +93,14 @@ def fakify(
+    ]
     symbolic_context = StatelessSymbolicContext(
-        dynamic_sizes=[DimDynamic.DYNAMIC] * n_dims,
+        dynamic_sizes=dynamic_sizes,
```
OK, trying to understand...so you start off with mostly STATIC here.

pianpwk replied:
Yeah, for now everything here can be static except those marked with Dim.AUTO. This doesn't interfere with the old Dim() specs.

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -136,10 +141,7 @@ def make_fake_inputs(
-        combined_args, dynamic_shapes
-    )
-    constraints = _process_dynamic_shapes(combined_args, transformed_dynamic_shapes)
+    constraints = _process_dynamic_shapes(combined_args, dynamic_shapes)
```
Where does AUTO come in?

pianpwk replied:
DimHints now get handled in `_process_dynamic_shapes`

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -559,7 +551,8 @@ def _export_to_torch_ir(
                     f,
-                    dynamic_shapes=transformed_dynamic_shapes,  # type: ignore[arg-type]
+                    dynamic_shapes=dynamic_shapes,  # type: ignore[arg-type]
+                    assume_static_by_default=True,
```
What does this buy?

pianpwk replied:
Nothing functionality wise, we just don't have to maintain this specs inverting which is already pretty confusing to me today

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -1011,20 +888,43 @@ def update_symbols(path, tensor, shape):
         def _create_static_dim(tensor, i, value):
             return _StaticDim(str(value), (int,), {"value": value})
 
+        # clean out decorators from user side, or previous export call
```
Interesting, so we'll only allow specs to dictate which ones are marked? Should we warn if any of these were set, asking to use specs instead?

pianpwk replied:
Ah added a comment in the last PR description line

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -1064,10 +964,10 @@ def _get_dim_name_mapping(
-        name_to_dim[dim.__name__] = dim
-        if isinstance(dim, _DerivedDim):
-            name_to_dim[dim.root.__name__] = dim.root  # type: ignore[attr-defined]
+        if isinstance(dim, _Dim):
```
`else` what? Maybe handle that explicitly with a pass and then assert / give a useful error if anything unexpected happens?

avikchaudhuri reviewed (approved):
Looks fine as long as tests pass, but I'm not entirely sure what the effect of the change is in this PR, could you write a summary please?

What I'd like is the some kind of invariant to hold around STATIC, constant dims, DYNAMIC, AUTO, and None here. Looks like we're not transforming specs any more, but using a marking scheme to get what we want. Who consumes these markings, is it just dynamo, or is it make_fx as well?

pianpwk commented:
> Looks fine as long as tests pass, but I'm not entirely sure what the effect of the change is in this PR, could you write a summary please?
> 
> What I'd like is the some kind of invariant to hold around STATIC, constant dims, DYNAMIC, AUTO, and None here. Looks like we're not transforming specs any more, but using a marking scheme to get what we want. Who consumes these markings, is it just dynamo, or is it make_fx as well?

Ah sorry, just realized the diff description didn't update the PR description too

pianpwk committed:
[export] simplify automatic dynamic shapes processing (#136591)

Summary:
Removing `_transform_shapes_for_default_dynamic` and `assume_static_by_default=False` as added in https://github.com/pytorch/pytorch/pull/133620.

This reverts back to `assume_static_by_default=True` with the use of dynamo decorators (e.g. `maybe_mark_dynamic, mark_static`, instead) for handling Dim.AUTO & Dim.STATIC instead. This is easier to maintain, as it doesn't requiring reasoning about "inverting" the dynamic_shapes specs, and also opens up usage of other decorators (`mark_dynamic, mark_unbacked`).

On the user side this change has no effect, but internally this means dynamic behavior is determined only by the `dynamic_shapes` specs (ignoring user-side input decorators following https://github.com/pytorch/pytorch/pull/135536), but transferring this information for _DimHints via decorators, for Dynamo/non-strict to create symbolic_contexts accordingly, e.g. https://github.com/pytorch/pytorch/blob/7c6d543a5b3d65d8f49420e60cda150faaa5b8a0/torch/_dynamo/variables/builder.py#L2646-L2666

One caveat is we don't raise errors for dynamic decorators on the user side, since we don't know if they're from user markings, or from re-exporting with inputs we've previously marked.


Differential Revision: D63358628

Pulled By: pianpwk

pianpwk commented:
@pytorchbot merge -f "unrelated lint"

desertfire commented:
> @pytorchbot merge -f "unrelated lint"

@pianpwk , as a retrospect, the `sam_fast` failure was really and shouldn't been force merged.

pianpwk commented:
> > @pytorchbot merge -f "unrelated lint"
> 
> @pianpwk , as a retrospect, the `sam_fast` failure was really and shouldn't been force merged.

@desertfire Ah sorry. I should've realized the test no longer timing out was related
